watchtower-plugin/src/retrier.rs

```diff
@@ -812,6 +1145,142 @@ mod tests {
+
+        {
+            tokio::time::sleep(Duration::from_secs(2)).await;
+            let state = wt_client.lock().unwrap();
+            assert!(state.get_retrier_status(&tower_id).unwrap().is_idle());
+            let tower = state.towers.get(&tower_id).unwrap();
+            assert_eq!(tower.status, TowerStatus::Unreachable);
+        }
+
+        let mut add_appointment_receipt = AppointmentReceipt::new(
+            cryptography::sign(&appointment.to_vec(), &wt_client.lock().unwrap().user_sk).unwrap(),
+            42,
+        );
+
+        // Mock a proper response
+        let server = MockServer::start();
+        add_appointment_receipt.sign(&tower_sk);
+        let add_appointment_response =
+            get_dummy_add_appointment_response(appointment.locator, &add_appointment_receipt);
+        let api_mock = server.mock(|when, then| {
+            when.method(POST).path("/add_appointment");
+            then.status(200)
+                .header("content-type", "application/json")
+                .json_body(json!(add_appointment_response));
+        });
+
+        // Patch the tower address
+        wt_client
+            .lock()
+            .unwrap()
+            .towers
+            .get_mut(&tower_id)
+            .unwrap()
+            .net_addr = server.base_url();
+
+        // Check pending data is still there now, and is it not once the retrier succeeds
+        assert_eq!(
+            wt_client
+                .lock()
+                .unwrap()
+                .towers
+                .get(&tower_id)
+                .unwrap()
+                .pending_appointments
+                .len(),
+            2,
+        );
+
+        // Send a retry flag to the retrier to force a retry.
+        tx.send((tower_id, RevocationData::None)).unwrap();
+
+        tokio::time::sleep(Duration::from_secs(MAX_ELAPSED_TIME as u64)).await;
+        // FIXME: Here we should be able to check this, however, due to httpmock limitations, we cannot return a response based on the request.
+        // Therefore, both requests will be responded with the same data. Given pending_appointments is a HashSet, we cannot even know which request
+        // will be sent first (sets are initialized with a random state, which decided the order or iteration).
+        // https://github.com/alexliesenfeld/httpmock/issues/49
+        // assert!(!wt_client.lock().unwrap().retriers.contains_key(&tower_id));
+        // assert!(wt_client
+        //     .lock()
+        //     .unwrap()
+        //     .towers
+        //     .get(&tower_id)
+        //     .unwrap()
+        //     .pending_appointments
+        //     .is_empty());
+
+        // This is not much tbh, but looks like its the best we can do at the moment without experiencing random errors.
+        // Depending on what appointment is sent first the api will be hit either one or two times.
+        assert!(api_mock.hits() >= 1 && api_mock.hits() <= 2);
+        task.abort();
```

Comment:
Q: What retrier error happen when the receipt doesn't match the appointment? Is it a signature error?

Reply:
`AddAppointmentError::SignatureError`